calc/vxlan_resolver.go

```diff
@@ -0,0 +1,476 @@
+		vxlanPools:                map[string]model.IPPool{},
+	}
+}
+
+func (c *VXLANResolver) RegisterWith(allUpdDispatcher *dispatcher.Dispatcher) {
+	allUpdDispatcher.Register(model.HostIPKey{}, c.OnHostIPUpdate)
+	allUpdDispatcher.Register(model.HostConfigKey{}, c.OnHostConfigUpdate)
+	allUpdDispatcher.Register(model.BlockKey{}, c.OnBlockUpdate)
+	allUpdDispatcher.Register(model.IPPoolKey{}, c.OnPoolUpdate)
+}
+
+func (c *VXLANResolver) OnBlockUpdate(update api.Update) (_ bool) {
+	// Update the routes map based on the provided block update.
+	key := update.Key.String()
+
+	deletes := set.New()
+	adds := set.New()
+	if update.Value != nil {
+		// Block has been created or updated.
+		// We don't allow multiple blocks with the same CIDR, so no need to check
+		// for duplicates here. Look at the routes contributed by this block and determine if we
+		// need to send any updates.
+		newRoutes := c.routesFromBlock(key, update.Value.(*model.AllocationBlock))
+		currentRoutes, ok := c.blockToRoutes[key]
+		if !ok {
+			currentRoutes = set.New()
+			c.blockToRoutes[key] = currentRoutes
+		}
+
+		for _, r := range newRoutes {
+			logCxt := logrus.WithField("newRoute", r)
+			if currentRoutes.Contains(r) {
+				logCxt.Debug("Desired VXLAN route already exists, skip")
+				continue
+			}
+
+			c.blockToRoutes[key].Add(r)
+			adds.Add(r)
+		}
+
+		currentRoutes.Iter(func(item interface{}) error {
+			r := item.(vxlanRoute)
+
+			// For each existing route which is no longer present, we need to delete it.
+			if _, ok := newRoutes[r.Key()]; ok {
+				// Exists, and we want it to - nothing to do.
+				return nil
+			}
+
+			// Current route is not in new set - we need to withdraw the route, and also
+			// remove it from internal state.
+			deletes.Add(r)
+			c.blockToRoutes[key].Discard(r)
+			return nil
+		})
+
+		// At this point we've determined the correct diff to perform based on the block update.
+		// Delete any routes which are gone for good, withdraw modified routes, and send updates for
+		// new ones.
+		deletes.Iter(func(item interface{}) error {
+			c.withdrawRoute(item.(vxlanRoute))
+			return nil
+		})
+		c.kickPendingRoutes(adds)
+	} else {
+		// Block has been deleted. Clean up routes that were contributed by this block.
+		routes := c.blockToRoutes[key]
+		routes.Iter(func(item interface{}) error {
+			c.withdrawRoute(item.(vxlanRoute))
+			return nil
+		})
+		delete(c.blockToRoutes, key)
+	}
+	return
+}
+
+// OnHostIPUpdate gets called whenever a node IP address changes. On an add/update,
+// we need to check if there are VTEPs or routes which are now valid, and trigger programming
+// of them to the data plane. On a delete, we need to withdraw any routes and VTEPs associated
+// with the node.
+func (c *VXLANResolver) OnHostIPUpdate(update api.Update) (_ bool) {
+	nodeName := update.Key.(model.HostIPKey).Hostname
+	logCxt := logrus.WithField("node", nodeName)
+	pendingSet, sentSet := c.routeSets()
```

Comment:
Actually generating these sets will make this logic O(n^2) to add n nodes.  I'd expect it to start getting sluggish at 1000-node scale and to explode by 10k node scale.

I had in mind to scan only the relevant part of the index for each type of update, if if you get a block update, you only need to look at the routes that were in the old version of that block since a block update can't trigger routes from other blocks to be suddenly valid.  If you get a node update, you only need to scan the routes that have that node as next hop.

The approach you've got looks correct though and should work fine for 100s of nodes, so I think any changes to be more precise in the indexing should be done in a follow-up increment.